### ISSUE TITLE
shared nodes

### ISSUE THREAD

**Issue by panleone** [panleone/NeuralNetwork#11]
let's say we have such an expression

```
y = BIN_OP(f(x), f(x)) 
y.eval()
```

right now the computation of `f(x)` is done twice. Find a way to compute only once.

This happens often, for example:

At the end of a neural network when computing the MSE loss
`loss =  (f(x) - target)*(f(x) - target)`

In a residual connection
`y = f(x) + g(f(x))`
